crates/data_components/src/s3_single_file_cached.rs

```diff
@@ -0,0 +1,458 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//! A wrapper around `ListingTable` for single S3 files that caches `ETag` and Version ID
+//! to avoid unnecessary re-scans when the file hasn't changed.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use arrow::datatypes::SchemaRef;
+use async_trait::async_trait;
+use datafusion::catalog::Session;
+use datafusion::datasource::listing::ListingTable;
+use datafusion::datasource::{TableProvider, TableType};
+use datafusion::error::Result as DataFusionResult;
+use datafusion::logical_expr::{Expr, TableProviderFilterPushDown};
+use datafusion::physical_plan::ExecutionPlan;
+use object_store::ObjectStore;
+use object_store::path::Path as ObjectStorePath;
+use tokio::sync::RwLock;
+
+use crate::refresh_skip::RefreshSkipTableProvider;
+
+/// Normalize an optional string by trimming whitespace and treating empty strings as None
+fn normalize_optional_string(s: Option<&String>) -> Option<String> {
+    s.and_then(|v| {
+        let trimmed = v.trim();
+        if trimmed.is_empty() {
+            None
+        } else {
+            Some(trimmed.to_string())
+        }
+    })
+}
+
+/// Check if two `ObjectMeta` represent the same file version based on version ID and `ETag`
+fn is_same_file_version(
+    cached: &object_store::ObjectMeta,
+    current: &object_store::ObjectMeta,
+) -> bool {
+    let cached_version = normalize_optional_string(cached.version.as_ref());
+    let current_version = normalize_optional_string(current.version.as_ref());
+    let cached_etag = normalize_optional_string(cached.e_tag.as_ref());
+    let current_etag = normalize_optional_string(current.e_tag.as_ref());
+
+    // Both version and etag are absent in both - no versioning info available, be conservative and assume different
+    if cached_version.is_none()
+        && current_version.is_none()
+        && cached_etag.is_none()
+        && current_etag.is_none()
+    {
+        return false;
+    }
+
+    // Check if version or etag presence differs (one has it, other doesn't) - different files
+    if (cached_version.is_some() != current_version.is_some())
+        || (cached_etag.is_some() != current_etag.is_some())
```

Comment:
This will return `false` when versions are the same, but etag is absent. Maybe this is intended though